zcash_primitives/src/sapling/keys.rs

```diff
@@ -157,6 +161,151 @@ impl FullViewingKey {
     }
 }
 
+/// The scope of a viewing key or address.
+///
+/// A "scope" narrows the visibility or usage to a level below "full".
+///
+/// Consistent usage of `Scope` enables the user to provide consistent views over a wallet
+/// to other people. For example, a user can give an external [`SaplingIvk`] to a merchant
+/// terminal, enabling it to only detect "real" transactions from customers and not
+/// internal transactions from the wallet.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Scope {
+    /// A scope used for wallet-external operations, namely deriving addresses to give to
+    /// other users in order to receive funds.
+    External,
+    /// A scope used for wallet-internal operations, such as creating change notes,
+    /// auto-shielding, and note management.
+    Internal,
+}
+
+/// A Sapling key that provides the capability to view incoming and outgoing transactions.
+///
+/// This key is useful anywhere you need to maintain accurate balance, but do not want the
+/// ability to spend funds (such as a view-only wallet).
+///
+/// It comprises the subset of the ZIP 32 extended full viewing key that is used for the
+/// Sapling item in a [ZIP 316 Unified Full Viewing Key][zip-0316-ufvk].
+///
+/// [zip-0316-ufvk]: https://zips.z.cash/zip-0316#encoding-of-unified-full-incoming-viewing-keys
+#[derive(Clone, Debug)]
+pub struct DiversifiableFullViewingKey {
+    fvk: FullViewingKey,
+    dk: zip32::DiversifierKey,
+}
+
+impl From<zip32::ExtendedFullViewingKey> for DiversifiableFullViewingKey {
+    fn from(extfvk: zip32::ExtendedFullViewingKey) -> Self {
+        Self {
+            fvk: extfvk.fvk,
+            dk: extfvk.dk,
+        }
+    }
+}
+
+impl DiversifiableFullViewingKey {
+    /// Parses a `DiversifiableFullViewingKey` from its raw byte encoding.
+    ///
+    /// Returns `None` if the bytes do not contain a valid encoding of a diversifiable
+    /// Sapling full viewing key.
+    pub fn from_bytes(bytes: &[u8; 128]) -> Option<Self> {
+        FullViewingKey::read(&bytes[..96]).ok().map(|fvk| Self {
+            fvk,
+            dk: zip32::DiversifierKey(bytes[96..].try_into().unwrap()),
+        })
+    }
+
+    /// Returns the raw encoding of this `DiversifiableFullViewingKey`.
+    pub fn to_bytes(&self) -> [u8; 128] {
+        let mut bytes = [0; 128];
+        self.fvk
+            .write(&mut bytes[..96])
+            .expect("slice should be the correct length");
+        bytes[96..].copy_from_slice(&self.dk.0);
+        bytes
+    }
+
+    /// Derives the internal `DiversifiableFullViewingKey` corresponding to `self` (which
+    /// is assumed here to be an external DFVK).
```

Comment:
It's a little bit error-prone that the type doesn't record whether this is an internal or external FVK, and so you can call `derive_internal` incorrectly on a `DiversifiableFullViewingKey` that is already internal. (OTOH, in most places internal and external keys are handled uniformly.)

Reply:
Note that this API does not expose a way to produce an internal `DiversifiableFullViewingKey`. This assumption is referring to the fact that we can't know whether `DiversifiableFullViewingKey::from_bytes` is parsing something that is external or internal (or not HD-derived at all), since that information is not stored in the encoding.